k2/csrc/fsa_algo.h

```diff
@@ -120,35 +122,40 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map = nullptr);
+   WfsaWithFbWeights.
+                                                                     // Caution:
+   we should probably use
+                                                                     // double
+   here; these kinds of algorithms
+                                                                     // are
+   extremely sensitive to roundoff for
+                                                                     // very
+   long FSAs. local_backpointers[ki] = -1  // will terminate a sequence..
          queue.push_back(ki)
          while (!queue.empty()) {
-            ji = queue.front()  // we have to be a bit careful about order here, to make sure
-                                // we always process states when they already have the
+            ji = queue.front()  // we have to be a bit careful about order here,
+   to make sure
+                                // we always process states when they already
+   have the
                                 // best cost they are going to get.  If
-                                // FSA was top-sorted at the start, which we assume, we could perhaps
-                                // process them in numerical order, e.g. using a heap.
-            queue.pop_front()
-            for each arc leaving state ji:
-                next_weight = local_forward_state_weights[ji] + arc_weights[this_arc_index]
-                if next_weight + backward_state_weights[arc_dest_state] < best_path_weight - beam:
-                    if arc label is epsilon:
-                        if next_weight < local_forward_state_weight[next_state]:
+                                // FSA was top-sorted at the start, which we
+   assume, we could perhaps
+                                // process them in numerical order, e.g. using a
+   heap. queue.pop_front() for each arc leaving state ji: next_weight =
+   local_forward_state_weights[ji] + arc_weights[this_arc_index] if next_weight
+   + backward_state_weights[arc_dest_state] < best_path_weight - beam: if arc
+   label is epsilon: if next_weight < local_forward_state_weight[next_state]:
                            local_forward_state_weight[next_state] = next_weight
                            local_backpointers[next_state] = ji
                 else:
                     add an arc to the output FSA, and create the appropriate
                     arc_map entry by following backpointers (hopefully you can
-                    figure out the details).  Note: the output FSA's weights can be
-                    computed later on, by calling code, using the info in arc_map.
+                    figure out the details).  Note: the output FSA's weights can
+   be computed later on, by calling code, using the info in arc_map.
  */
 void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
-                         std::vector<std::vector<int32_t>> *arc_map);
+                         std::vector<float> *arc_weights_b);
```

Comment:
Can I change the interface to this? Is `arc_map` needed by following steps?

Reply:
No!   The whole point of this framework is to add differentiability.
You'll see why once we do the pytorch integration.

Reply:
Read notes/python.txt to get some feel for how this will interact with PyTorch, that may make it more clear why the interface needs to be like this.

Reply:
OK, then I'll revert it back to the interface, and add the arc_maps computation